src/datepicker/datepicker-tools.ts

```diff
@@ -87,37 +87,68 @@ export function buildMonths(
     calendar: NgbCalendar, date: NgbDate, state: DatepickerViewModel, i18n: NgbDatepickerI18n,
     force: boolean): MonthViewModel[] {
   const {displayMonths, months} = state;
-  const newMonths = [];
+  const newFirstMonthStart = calendar.getNext(date, 'm', 0);
```

Comment:
What about the following implementation?
Not sure it's more/less optimised, but it is shorter and puts less strain on my brain :)
(all your tests pass as before)

```ts
  // move old months to a temporary array
  const monthsToReuse = months.splice(0, months.length);

  // generate new first dates, nullify or reuse months
  const firstDates = Array.from({length: displayMonths}, (_, i) => {
    const firstDate = calendar.getNext(date, 'm', i);
    months[i] = null;

    if (!force) {
      const reusedIndex = monthsToReuse.findIndex(month => month.firstDate.equals(firstDate));
      // move reused month back to months
      if (reusedIndex !== -1) {
        months[i] = monthsToReuse.splice(reusedIndex, 1)[0];
      }
    }

    return firstDate;
  });

  // rebuild nullified months
  firstDates.forEach((newDate, i) => {
    if (months[i] === null) {
      months[i] = buildMonth(calendar, newDate, state, i18n, monthsToReuse.shift() || {} as MonthViewModel);
    }
  });
```